datasette/app.py

```diff
@@ -1450,20 +1228,88 @@ async def allowed(
         Uses SQL to check permission for a single resource without fetching all resources.
         This is efficient - it does NOT call allowed_resources() and check membership.
 
+        If resource is not provided, defaults to InstanceResource() for instance-level actions.
+
         Example:
             from datasette.resources import TableResource
             can_view = await datasette.allowed(
                 action="view-table",
                 resource=TableResource(database="analytics", table="users"),
                 actor=actor
             )
+
+            # For instance-level actions, resource can be omitted:
+            can_debug = await datasette.allowed(action="permissions-debug", actor=actor)
         """
         from datasette.utils.actions_sql import check_permission_for_resource
+        from datasette.resources import InstanceResource
+        import datetime
+
+        if resource is None:
+            resource = InstanceResource()
+
+        # Check if this action has also_requires - if so, check that action first
+        action_obj = self.actions.get(action)
+        if action_obj and action_obj.also_requires:
+            # Must have the required action first
+            if not await self.allowed(
+                action=action_obj.also_requires,
+                resource=resource,
+                actor=actor,
+            ):
+                return False
 
-        return await check_permission_for_resource(
-            self, actor, action, resource.parent, resource.child
+        result = await check_permission_for_resource(
+            datasette=self,
+            actor=actor,
+            action=action,
+            parent=resource.parent,
+            child=resource.child,
         )
 
+        # Check actor restrictions after SQL permissions
+        # If the SQL check says "yes" but actor has restrictions, verify action is allowed
+        if result and actor and "_r" in actor:
+            from datasette.default_permissions import restrictions_allow_action
+
+            # Convert Resource to old-style format for restrictions check
+            if resource.parent and resource.child:
+                old_style_resource = (resource.parent, resource.child)
+            elif resource.parent:
+                old_style_resource = resource.parent
+            else:
+                old_style_resource = None
+
+            # If restrictions don't allow this action, deny it
+            if not restrictions_allow_action(
+                self, actor["_r"], action, old_style_resource
+            ):
+                result = False
+
+        # Log the permission check for debugging
```

Comment:
Instead of storing a `"resource"` key, lets store a `"parent"` and `"child"` key, both of which are string or None. In fact lets start using a dataclass for the things that go in `self._permission_checks` just so we get types in there.